### PR TITLE
Refactor probe parsing

### PR DESCRIPTION
**Pull request by gildub**
Reduces usage by triggering probes only on a per demand need.
Also prepares for moving the probes to their respective providers or to
proper specific module such as MiqIPMI or Host/Helper.

Factorizes the log encapsulating each probe.

### CODE REVIEW

**Comment by Fryguy** [comment]
Very cool!  I'll need to take a deeper look, but I like the direction this is going!

**Comment by gildub** [comment]
@Fryguy, thanks

File: lib/manageiq-network_discovery.rb [hunk 1]
```diff
@@ -1 +1,19 @@
...
+    autoload :MSScvmmProbe, 'manageiq/network_discovery/modules/MSScvmmProbe'
+    autoload :MSVirtualServerProbe, 'manageiq/network_discovery/modules/MSVirtualServerProbe'
+    autoload :RedHatRhevmProbe, 'manageiq/network_discovery/modules/RedHatRhevmProbe'
+    autoload :VMwareEsxVcProbe, 'manageiq/network_discovery/modules/VMwareEsxVcProbe'
+    autoload :VMwareServerProbe, 'manageiq/network_discovery/modules/VMwareServerProbe'
+    autoload :WindowsProbe, 'manageiq/network_discovery/modules/WindowsProbe'
+  end
+
+  module Providers
+    module Openstack
```
Comment (Fryguy) [inline lib/manageiq-network_discovery.rb:13]: I assume this is temporary for extraction purposes?

**Comment by Fryguy** [comment]
LGTM, but I want at least one more set of eyes on this... @bdunne @roliveri ?

**Comment by Fryguy** [comment]
cc @jrafanie

**Review by Fryguy — APPROVED** [review]

**Comment by blomquisg** [comment]
I really like this idea.  I'd like to refine it a bit over time (maybe not for this PR).

Basically, I wonder if it's better to move the network probes to the providers and allow them to supply/register their own probes.

Sometimes there's startup/initializer issues that make that hard to do.  So, it's fine if that's just not a good idea.  But, I'd like to figure out how much of this we could push down to the individual providers so they control the logic of how to do the network probe.

Nice work @gildub!

**Comment by Fryguy** [comment]
> Basically, I wonder if it's better to move the network probes to the providers and allow them to supply/register their own probes.

That's the direction @gildub is going with the openstack probe. See https://github.com/ManageIQ/manageiq-providers-openstack/pull/115

File: lib/manageiq-network_discovery.rb (hunk 1 above)
Comment (gildub) [reply lib/manageiq-network_discovery.rb:13]: Yes each autoload to go to their respective providers

**Comment by gildub** [comment]
@blomquisg, @Fryguy, 
Thanks for the feed back!
Could we get it merged? There isn't any impact, just moving the code around. From that I'm happy to relocate the different probes to their respective providers.

File: lib/manageiq/network_discovery.rb [hunk 2]
```diff
@@ -79,11 +91,20 @@ def self.scanHost(sInfo)
...
+      rescue Timeout::Error
+        pingOk = false
+      end
+
+      if pingOk
+        # Trigger probes
+        sInfo.discover_types.each do |type|
+          klass = Object.const_get(PROVIDERS_BY_TYPE[type])
+          $log.info "#{klass}: probing ip = #{sInfo.ipaddr}" if $log
+          klass.send(:probe, sInfo)
```
Comment (Fryguy) [inline lib/manageiq/network_discovery.rb:104]: Minor, but this can just be `klass.probe(sInfo)`

Comment (gildub) [reply lib/manageiq/network_discovery.rb:104]: I missed that. Thanks anyway!

**Pull request merged**